### PR TITLE
Rails Group Project - Recipe App

### PR DESCRIPTION
**Pull request by kwesiObuobi**
### Layout
- We followed the layout of the wireframes provided

### Authentication
- Created a Login page and Registration page
- Built them with `Devise`

### App Pages
- Food List page
  - displayed a list of food added by the logged-in user as in the wireframe (displays also quantity of a given food).
  - led to a form that allows users to add new food.

- Recipe List page
  - displayed a list of recipes created by the logged-in user as in the wireframe.
  - led to recipe details.
  - allowed the user to delete if they are the owner

- Public Recipe list
  - displayed a list of all public recipes ordered by the newest as in the wireframe.
  - led to recipe details.

- Recipe Details page
  - displayed a toggle button that allows for a recipe to be made public or private.
  - displayed recipe details if the recipe is `public` or the user is the owner of the recipe
  - led to a form that allows the user to add new food if the user is the owner of the recipe

- General Shopping List page
  - showed the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
  - counted the total food items and the total price of the missing food.

- Created a navigation menu that allows users to open all of the pages we created.

### Database Queries
- Ensured that there are no `N + 1 queries` happening

### Tests
- Wrote unit tests for our models
- Wrote integration tests using the `capybara` gem

### CODE REVIEW

**Review by HAMMAS-SALEEM — CHANGES_REQUESTED** [review]
# Required Changes :recycle:
Ukhv

## Hey @kwesiObuobi, @yegonkimutai :wave:

## Highlights :1st_place_medal:

✔️ Descriptive PR Title
✔️ Descriptive PR Summary
✔️ Following Gitflow
✔️ Following some requirements
✔️ Great work on making the tests 👍
✔️ Great work on making the associations of far 😊
✔️ Great work on building this project collaboratively 🔥

Good job on implementing some of the requirements, there are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/recipes/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+<h1><%= @recipe.name%></h1>
```
[the remark this answers is not in the document]
Comment (kwesiObuobi) [reply app/views/recipes/show.html.erb:1]: Thanks, @HAMMAS-SALEEM 
Changes have been made now

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    else
+      render :new, status: 422
+    end
+  end
+
+  def destroy
+    @recipe = Recipe.find(params[:id])
+
+    if @recipe.destroy
+      redirect_to recipes_path, notice: 'Recipe was successfully deleted.'
```
[the remark this answers is not in the document]
Comment (kwesiObuobi) [reply app/controllers/recipes_controller.rb:33]: Thanks for flagging this. It's fixed now.

File: app/views/recipes_foods/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+<h1>Add Ingredient for <%= @recipe_name %></h1>
+<div class="d-flex flex-column mb-3">
+<div class="col">
+<%= form.label :quantity %>
+<%= form.number_field :quantity %>
+</div>
+<div class="col">
+<%= form.label :food_id%>
+<%= form.select :food_id, Food.all.map {|f| [f.name, f.id]} %>
+</div>
```
[the remark this answers is not in the document]
Comment (kwesiObuobi) [reply app/views/recipes_foods/new.html.erb:22]: Thanks, @HAMMAS-SALEEM 
We have solved it now.

File: app/views/devise/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<h2 class='signup-title'>Log in</h2>
```
[the remark this answers is not in the document]
Comment (kwesiObuobi) [reply app/views/devise/sessions/new.html.erb:1]: Thanks
Gem is added now.

File: README.md [hunk 5]
```diff
@@ -1,24 +1,162 @@
-# README
+<a name="readme-top"></a>
```
[the remark this answers is not in the document]
Comment (kwesiObuobi) [reply README.md:2]: Ok sir
We will consider that.

**Review by nahnahdev — APPROVED** [review]
## Project Status: APPROVED :green_circle: 
Hi @kwesiObuobi and @yegonkimutai , 

Great Job implementing all the Previous reviewer's suggestions :clap: Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights :fire: 
- Titles centered as expected :heavy_check_mark: 
- Authorization issues solved :heavy_check_mark: 
- Deleting a recipe that has some food items and food that is already being used in a recipe works without errors :heavy_check_mark: 
- Shopping list bug solved :heavy_check_mark: 
- Confirmation of email address implemented :heavy_check_mark: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**